fix ironic sync with unset capabilities

msherman64 opened the pull request:
Addresses issue #62

diurnalist commented on doni/driver/worker/ironic.py:
```diff
@@ -216,12 +216,15 @@ def process(
+                    for key, value in hw_props.get("baremetal_capabilities").items()
+                ]
+            )
+        except AttributeError:
```
This is a bit odd to me, to catch the error if `items()` doesn't exist. What about this
```python
hw_capabilities = hw_props.get("baremetal_capabilities")
if hw_capabilities:
    capabilities_string = ",".join([f"{key}:{value}" for key, value in hw_capabilities.items()])
else:
    capabilities_string = None
```

msherman64 replied:
Added a check for being a Mapping, since otherwise the type might not have ".items()"

I'm always a bit fuzzy on the most "pythonic" way to handle this stuff

diurnalist replied:
I think if you know the shape of the thing, it's fine to just to truthy/falsy. Only when the inputs are really variable does it make sense to catch low-level errors, IMO. Like this:

```python
def try_int(num):
    try:
        return int(num)
    except ValueError:
        return None
```

diurnalist reviewed (approved):
🚢

diurnalist commented on doni/driver/worker/ironic.py:
```diff
@@ -215,13 +215,14 @@ def process(
-                for key, value in hw_props.get("baremetal_capabilities").items()
-            ]
-        )
+        if isinstance(hw_capabilities, Mapping):
```
almost... the `typing` library is used for type hinting, aka things that shouldn't actually affect or get called at runtime. I'm surprised that this actually evaluates to True properly, maybe there's some odd reason for this, or they have coded it in anticipation or in reaction to it being called like this. I think `isinstance(hw_capabilities, dict)` is sufficient, but what if it's empty? Then the capabilities string will be the empty string, is that OK?

msherman64 replied:
good catch. I'd meant to reference `collections.abc.Mapping` there. Technically this is coming from a FlexibleDict field, so maybe I don't need to check type here? Wasn't sure if that would always return a `dict` vs an OrderedDict or something else.

msherman64 committed:
Handle case where hw_capabilities is not a Mapping

This method should return either None, or a valid string.

Since we call .items(), we need to check for two cases:
1. hw_capabilities is not a Mapping (cannot call .items)
2. hw_capabilites is an empty Mapping (join would return empty string)

abc.Mapping was chosen instead of dict to support other implementations.
(aka, I'm not sure exactly what a FlexibleDictField might hold)
General advice seems to indicate against comparing directly to the dict
class. e.g. https://stackoverflow.com/a/62430764

msherman64 committed:
Update doni/driver/worker/ironic.py

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
